File: app-kotlin/src/main/java/com/TritiumGaming/phasmophobiaevidencepicker/data/persistent/theming/AThemeControl.java
package com.TritiumGaming.phasmophobiaevidencepicker.data.persistent.theming;

import android.annotation.SuppressLint;
import android.content.Context;
import android.util.Log;

import androidx.annotation.StyleRes;

import com.TritiumGaming.phasmophobiaevidencepicker.data.viewmodels.GlobalPreferencesViewModel;

import java.util.ArrayList;

/**
 * ColorSpaceData class
 *
 * @author TritiumGamingStudios
 */
public abstract class AThemeControl {

    protected @StyleRes int defaultStyle;

    protected ArrayList<CustomTheme> themes = new ArrayList<>();
    protected int savedIndex, selectedIndex;

    public AThemeControl(Context context) {
        build(context);
    }

    public void init(int savedIndex) {
        setSavedIndex(savedIndex);
        setSelectedIndex(getSavedIndex());
    }

    @SuppressLint("Recycle, ResourceType")
    protected abstract void build(Context context);

    public void setSavedIndex() {
        this.savedIndex = selectedIndex;
    }

    public void setSavedIndex(int savedIndex) {
        this.savedIndex = savedIndex < themes.size() ? savedIndex : 0;
    }

    /**
     * @param selectedIndex
     */
    public void setSelectedIndex(int selectedIndex) {
        this.selectedIndex = selectedIndex;
    }

    /**
     * @return
     */
    public int getSavedIndex() {
        return savedIndex;
    }

    public void revertSelection() {
        selectedIndex = savedIndex;
    }

    /**
     * @return
     */
    public int getSelectedIndex() {
        return selectedIndex;
    }

    /**
     * @param dir
     */
    public void iterateSelection(int dir) {

        selectedIndex += dir;

        if (selectedIndex < 0) {
            selectedIndex = themes.size() - 1;
        }
        else if (selectedIndex >= themes.size()) {
            selectedIndex = 0;
        }
    }

    public CustomTheme getCurrentTheme() {

        CustomTheme theme = themes.get(selectedIndex);
        if(theme == null) {
            new CustomTheme(0, -1, defaultStyle);
        }

        return theme;

    }

    /**
     * @return
     */
    public int getCurrentName() {
        return getCurrentTheme().getName();
    }

    /**
     * @return
     */
    public @StyleRes int getCurrentStyle() {
        return getCurrentTheme().getStyle();
    }

    public CustomTheme getAppThemeAt(int index) {

        CustomTheme tempTheme = new CustomTheme(0, -1, defaultStyle);

        if(index >= themes.size())
        {
            return tempTheme;
        }

        CustomTheme theme = themes.get(index);
        if(theme == null) {
            return tempTheme;
        }

        return theme;

    }

    public @StyleRes int getThemeAt(int index) {

        CustomTheme tempTheme = new CustomTheme(0, -1, defaultStyle);

        if(index >= themes.size())
        {
            return tempTheme.getStyle();
        }

        CustomTheme theme = themes.get(index);
        if(theme == null) {
            return tempTheme.getStyle();
        }

        return theme.getStyle();

    }

}

File: app-legacy/src/main/java/com/tritiumgaming/phasmophobiaevidencepicker/presentation/viewmodel/controllers/theming/AThemeControl.kt
package com.tritiumgaming.phasmophobiaevidencepicker.presentation.viewmodel.controllers.theming

import android.content.Context
import androidx.annotation.StyleRes
import com.tritiumgaming.phasmophobiaevidencepicker.domain.model.settings.ThemeModel

abstract class AThemeControl(context: Context) {
    @StyleRes
    var defaultStyle: Int = 0

    protected var themes: ArrayList<ThemeModel> = ArrayList()

    private var savedIndex: Int = 0
    var selectedIndex: Int = 0
        set(value) {
            var tempIndex = value
            if (tempIndex < 0 || tempIndex >= themes.size) {
                tempIndex = getIndexOfID(ThemeModel.defaultTheme.iD)
            }

            field = tempIndex
        }

    val currentTheme: ThemeModel
        get() {
            if (selectedIndex >= themes.size || selectedIndex < 0) {
                return ThemeModel.defaultTheme
            }

            val theme = themes[selectedIndex]

            return theme
        }

    val currentName: Int
        get() = currentTheme.name

    val iD: String
        get() = currentTheme.iD ?: ThemeModel.defaultTheme.iD ?: ""

    protected abstract fun build(context: Context)

    init {
        build(context)
    }

    fun init(savedID: String) {
        saveIndex(getIndexOfID(savedID))
        selectedIndex  = savedIndex
    }

    private fun getIndexOfID(savedID: String?): Int {
        for (i in themes.indices) {
            if (themes[i].iD == savedID) {
                return i
            }
        }

        return 0
    }

    fun saveSelectedIndex() {
        this.savedIndex = selectedIndex
    }

    fun saveIndex(index: Int) {
        var savedIndex = index
        if (savedIndex < 0 || savedIndex >= themes.size) {
            savedIndex = getIndexOfID(ThemeModel.defaultTheme.iD)
        }

        this.savedIndex = savedIndex
    }

    fun revertToSavedIndex() {
        selectedIndex = savedIndex
    }

    @JvmOverloads
    fun iterateSelectedIndex(dir: Int, start: Int = selectedIndex) {
        var currentIndex = selectedIndex
        currentIndex += dir

        if (currentIndex < 0) {
            currentIndex = themes.size - 1
        } else if (currentIndex >= themes.size) {
            currentIndex = 0
        }
        selectedIndex = currentIndex

        if (!currentTheme.isUnlocked) {
            if (start != selectedIndex) {
                iterateSelectedIndex(dir, start)
            }
        }
    }

    fun getThemeAtIndex(index: Int): ThemeModel {
        if (index >= themes.size || selectedIndex < 0) {
            return ThemeModel.defaultTheme
        }

        val theme = themes[index]

        return theme
    }

    fun getIndexOfTheme(theme: ThemeModel): Int {
        for (i in 0..<themes.size) {
            if (themes[i].iD == theme.iD) {
                return i
            }
        }
        return 0
    }

    fun getThemeByUUID(uuid: String): ThemeModel {
        for (customTheme in themes) {
            if (customTheme.iD == uuid) {
                return customTheme
            }
        }
        return ThemeModel.defaultTheme
    }

    fun revertAllUnlockedStatuses() {
        for (theme in themes) {
            theme.revertUnlockStatus()
        }
    }
}
